packages/api-plugin-carts/src/util/updateCartFulfillmentGroups.js

```diff
@@ -1,61 +1,124 @@
+ * @param {Object[]} currentGroups The current cart fulfillment groups array
+ * @param {String} fulfillmentType Specific fulfillment type to be checked
+ * @param {Object} item The item (product) object
+ * @returns {undefined}
+ */
+function checkAndAddToGroup(currentGroups, fulfillmentType, item) {
+  let removeFromGroups = determineRemoveFromGroups(currentGroups, item._id, item.shopId);
+  if (removeFromGroups && removeFromGroups.length > 0) {
+    removeFromGroups = (removeFromGroups || []).map((group) => {
+      group.itemIds = (group.itemIds || []).filter((itemId) => item._id !== itemId);
+      return group;
+    });
+  }
+
+  const addToGroup = determineAddToGroup(currentGroups, fulfillmentType, item.shopId);
+  if (!addToGroup) {
+    // If no compatible group, add one with initially just this item in it
+    currentGroups.push({
+      _id: Random.id(),
+      itemIds: [item._id],
+      shopId: item.shopId,
+      type: fulfillmentType
+    });
+  } else if (!addToGroup.itemIds) {
+    // If there is a compatible group but it has no items array, add one with just this item in it
+    addToGroup.itemIds = [item._id];
+  } else if (!addToGroup.itemIds.includes(item._id)) {
+    // If there is a compatible group with an items array but it is missing this item, add this item ID to the array
+    addToGroup.itemIds.push(item._id);
+  }
+}
+
 /**
  * @summary Updates the `shipping` property on a `cart`
  * @param {Object} context App context
  * @param {Object} cart The cart, to be mutated
  * @returns {undefined}
  */
-export default function updateCartFulfillmentGroups(context, cart) {
+export default async function updateCartFulfillmentGroups(context, cart) {
   // Every time the cart is updated, create any missing fulfillment groups as necessary.
   // We need one group per type per shop, containing only the items from that shop.
   // Also make sure that every item is assigned to a fulfillment group.
+  const { collections } = context;
+  const { Fulfillment } = collections;
   const currentGroups = cart.shipping || [];
 
-  (cart.items || []).forEach((item) => {
-    let { supportedFulfillmentTypes } = item;
+  for (const item of (cart.items || [])) {
+    const { supportedFulfillmentTypes } = item;
+
+    // This is a new optional field that UI can pass in case the user selects fulfillment type
+    // for each item in the product details page instead of waiting till checkout
+    let { selectedFulfillmentType } = item;
+
     // Do not re-allocate the item if it is already in the group. Otherwise difficult for other code
     // to create and manage fulfillment groups
-    const itemAlreadyInTheGroup = currentGroups.find(({ itemIds }) => itemIds && itemIds.includes(item._id));
-    if (itemAlreadyInTheGroup) return;
+    // Commenting out the below check since the item by default (if no fulfillment type is provided) will be assigned to 'undecided' group
```

Comment:
These comments probably belong more in a PR comment than in the actual code. Once this PR is merged the context is lost.

Reply:
Updated the comments in code